trillian/ctfe/storage/storage.go

```diff
@@ -0,0 +1,17 @@
+package storage
+
+import (
+	"context"
+)
+
+// NonLeafCertificateChainStorage is an interface which allows CTFE binaries to use different storage implementations for non-leaf certificate chains.
+type NonLeafCertificateChainStorage interface {
+	// FindAll returns all key-value pairs of non-leaf certificate chains.
+	FindAll(ctx context.Context) (map[string][]byte, error)
+
+	// FindByHash returns the non-leaf certificate chain associated with the provided hash.
+	FindByHash(ctx context.Context, hash string) ([]byte, error)
+
+	// Add inserts the key-value pair of non-leaf certificate chain.
+	Add(ctx context.Context, hash string, chain []byte) error
```

Comment:
Thinking about this transactionally, I would expect this to take a whole chain. This could open a transaction internally and then ensure that all elements of the chain are stored, or none are. This is probably more performant, easier to reason about, and easier for client code to call. Thoughts?

Reply:
Can you elaborate more on this? This method takes the chain without the leaf certificate, so all elements of the chain is already presented as `[]byte`. The current design does not take the database transaction into consideration. If the chain is failed to be stored in CTFE storage, it will not proceed to call the Trillian gRPC API.

Reply:
Ah, the whole chain is stored in the `[]byte` ? I naively expected that a chain would be a `[][]byte`, and so `[]byte` was a single cert in the chain.

As long as it's clear to everyone how to parse the individual certs from the concatenated bytes, then this comment can be closed.

Reply:
What if the `IssuanceChainService.Add` takes `[]ct.ASN1Cert` and convert them into bytes before calling the `Add` method in the storage layer?

Reply:
Sure, that works. But what if someone else wants to use this interface to read them? How do they convert the `[]byte` back to `[]ct.ASN1Cert` ? It seems like the `IssuanceChainStorage` interface isn't doing its job if a user of it can't get the Issuance Chains back out of it.

There are 3 options I would have considered for writing this:
 1. Make the interface IssuanceChain specific, and then make the types passed in/out be `[]byte` for hashes, and `[]ct.ASN1Cert` for cert chains. How these are encoded/decoded is completely encapsulated within the implementations of the interface, and clients only deal with the objects that they expect.
 2. Make the interface IssuanceChain specific, and make the types be `string` for hashes, and `[]byte` for chains. The interface must document how these are to be encoded/decoded
 3. Make the interface completely generic, and simply represent a key-value store. Then my types are `string` for key, and `[]byte` for values. The storage layer then has no opinion or pretense that what it returns must be certs chains.

In order of preference, I would rank these [1, 3, 2].

Reply:
Thinking about it more, I actually prefer 3 the most. Unless the implementations need to care about certs or hashes, then a simple `string -> []byte` map logical storage is nice and clean.

Reply:
The storage is indeed a key-value store for issuance chain. The implementation does not need to care about the hash and cert as the conversion will be done in the issuance chain service layer. 

Replaced `hash` with `key` as the field name. Let me know if you would like the `chain` to be renamed as `value`.